### PR TITLE
Remove two variables from BlockContents class and don't use class Block for compressed block

### PR DESCRIPTION
**Pull request by siying**
Summary: We carry compression type and "cachable" variables for every block in the block cache, while they take well-known values. 8-byte is wasted for each block (2-byte for useful information but it takes 8 bytes because of padding). With this change, these two variables are removed.

The cachable information is only useful in the process of reading the block. We use other information to infer from it. For compressed blocks, the compression type is a part of the block content itself so we can get it from there.

Some code is slightly refactored so that the cachable information can flow better.

Another change is to only use class BlockContents for compressed block, and narrow the class Block to only be used for uncompressed blocks. This can make the Block class less confusing. It also saves tens of bytes for each block in compressed block cache.

Test Plan: Run tests with ASAN

### CODE REVIEW

File: table/block.cc [hunk 1]
```diff
@@ -781,47 +781,45 @@ Block::Block(BlockContents&& contents, SequenceNumber _global_seqno,
...
+        restart_offset_ = map_offset - num_restarts_ * sizeof(uint32_t);
+
+        if (restart_offset_ > map_offset) {
+          // map_offset is too small for NumRestarts() and
+          // therefore restart_offset_ wrapped around.
+          size_ = 0;
           break;
-        default:
-          size_ = 0;  // Error marker
-      }
```
Comment (siying) [inline table/block.cc]: These lines didn't change. The only change is that `if (compression_type() == kNoCompression)` is removed. But "make format" changed all those lines.

**Comment by siying** [comment]
The appveyor failure is not related.

File: table/block_based_table_reader.cc [hunk 2]
```diff
@@ -115,6 +114,11 @@ void DeleteCachedEntry(const Slice& /*key*/, void* value) {
   delete entry;
 }
 
+void DeleteBlockContents(const Slice& /*key*/, void* value) {
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: What was the problem with DeleteCachedEntry?

File: table/block_based_table_reader.cc [hunk 3]
```diff
@@ -253,7 +257,8 @@ class PartitionIndexReader : public IndexReader, public Cleanable {
               index_value_is_full_),
           index_block_->NewIterator<IndexBlockIter>(
               icomparator_, icomparator_->user_comparator(), nullptr,
-              kNullStats, true, index_key_includes_seq_, index_value_is_full_));
+              kNullStats, true, index_key_includes_seq_, index_value_is_full_,
+              false /* block_contents_pinned */));
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Is the default value false? If yes, should we skip passing false?

File: table/block_based_table_reader.cc [hunk 4]
```diff
@@ -262,7 +267,8 @@ class PartitionIndexReader : public IndexReader, public Cleanable {
           table_, ro, *icomparator_,
           index_block_->NewIterator<IndexBlockIter>(
               icomparator_, icomparator_->user_comparator(), nullptr,
-              kNullStats, true, index_key_includes_seq_, index_value_is_full_),
+              kNullStats, true, index_key_includes_seq_, index_value_is_full_,
+              false /* block_contents_pinned */),
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Is the default value false? If yes, should we skip passing false?

File: table/block_based_table_reader.cc [hunk 5]
```diff
@@ -280,7 +286,8 @@ class PartitionIndexReader : public IndexReader, public Cleanable {
     Statistics* kNullStats = nullptr;
     index_block_->NewIterator<IndexBlockIter>(
         icomparator_, icomparator_->user_comparator(), &biter, kNullStats, true,
-        index_key_includes_seq_, index_value_is_full_);
+        index_key_includes_seq_, index_value_is_full_,
+        false /* block_contents_pinned */);
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Is the default value false? If yes, should we skip passing false?

File: table/block_based_table_reader.cc [hunk 6]
```diff
@@ -545,7 +552,7 @@ class HashIndexReader : public IndexReader {
     return index_block_->NewIterator<IndexBlockIter>(
         icomparator_, icomparator_->user_comparator(), iter, kNullStats,
         total_order_seek, index_key_includes_seq_, index_value_is_full_,
-        prefix_index_.get());
+        false /* block_contents_pinned */, prefix_index_.get());
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Can we add an inline comment that for indexes there is no value in having them pinned so we always set it to false.

File: table/block_based_table_reader.cc [hunk 7]
```diff
@@ -1230,32 +1235,35 @@ Status BlockBasedTable::GetDataBlockFromCache(
 
   // found compressed block
   RecordTick(statistics, BLOCK_CACHE_COMPRESSED_HIT);
-  compressed_block = reinterpret_cast<Block*>(
+  compressed_block = reinterpret_cast<BlockContents*>(
       block_cache_compressed->Value(block_cache_compressed_handle));
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: So block_cache_compressed will store BlockContent instead of Block? This would be a major change? Where is that we are documenting this change?

File: table/block_based_table_reader.cc [hunk 8]
```diff
@@ -1230,32 +1235,35 @@ Status BlockBasedTable::GetDataBlockFromCache(
...
 
   // Retrieve the uncompressed contents into a new buffer
   BlockContents contents;
-  UncompressionContext uncompresssion_ctx(compressed_block->compression_type(),
-                                          compression_dict);
-  s = UncompressBlockContents(uncompresssion_ctx, compressed_block->data(),
-                              compressed_block->size(), &contents,
-                              format_version, ioptions, allocator);
+  UncompressionContext uncompresssion_ctx(compression_type, compression_dict);
+  s = UncompressBlockContents(uncompresssion_ctx, compressed_block->data.data(),
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Should we change UncompressBlockContents to receive Slice instead of data and size separately?

File: table/block_based_table_reader.cc [hunk 9]
```diff
@@ -1230,32 +1235,35 @@ Status BlockBasedTable::GetDataBlockFromCache(
...
+                              compressed_block->data.size(), &contents,
+                              rep->table_options.format_version, rep->ioptions,
+                              GetMemoryAllocator(rep->table_options));
 
   // Insert uncompressed block into block cache
   if (s.ok()) {
     block->value =
-        new Block(std::move(contents), compressed_block->global_seqno(),
+        new Block(std::move(contents), rep->get_global_seqno(is_index),
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: How come is_index was not relevant before?

File: table/block_based_table_reader.cc [hunk 10]
```diff
@@ -1300,63 +1308,69 @@ Status BlockBasedTable::PutDataBlockToCache(
     const Slice& block_cache_key, const Slice& compressed_block_cache_key,
     Cache* block_cache, Cache* block_cache_compressed,
     const ReadOptions& /*read_options*/, const ImmutableCFOptions& ioptions,
-    CachableEntry<Block>* block, Block* raw_block, uint32_t format_version,
-    const Slice& compression_dict, size_t read_amp_bytes_per_bit, bool is_index,
-    Cache::Priority priority, GetContext* get_context,
-    MemoryAllocator* allocator) {
-  assert(raw_block->compression_type() == kNoCompression ||
+    CachableEntry<Block>* block, BlockContents* raw_block_contents,
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Why did we change from Block to BlockContent here?

File: table/block_based_table_reader.cc [hunk 11]
```diff
@@ -1300,63 +1308,69 @@ Status BlockBasedTable::PutDataBlockToCache(
...
-    s = UncompressBlockContents(uncompression_ctx, raw_block->data(),
-                                raw_block->size(), &contents, format_version,
-                                ioptions, allocator);
+    s = UncompressBlockContents(
+        uncompression_ctx, raw_block_contents->data.data(),
+        raw_block_contents->data.size(), &uncompressed_block_contents,
+        format_version, ioptions, allocator);
   }
   if (!s.ok()) {
-    delete raw_block;
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: We used to delete raw_block? Why do not we need to delete raw_block_contents?

File: table/block_based_table_reader.cc [hunk 12]
```diff
@@ -1300,63 +1308,69 @@ Status BlockBasedTable::PutDataBlockToCache(
...
 
-  if (raw_block->compression_type() != kNoCompression) {
-    block->value = new Block(std::move(contents), raw_block->global_seqno(),
+  if (raw_block_comp_type != kNoCompression) {
+    block->value = new Block(std::move(uncompressed_block_contents), seq_no,
                              read_amp_bytes_per_bit,
                              statistics);  // uncompressed block
   } else {
-    block->value = raw_block;
-    raw_block = nullptr;
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: We used to set raw_block to nullptr? Do we need to do the same with raw_block_contents?

File: table/block_based_table_reader.cc [hunk 13]
```diff
@@ -1300,63 +1308,69 @@ Status BlockBasedTable::PutDataBlockToCache(
...
   }
   if (!s.ok()) {
-    delete raw_block;
     return s;
   }
 
-  if (raw_block->compression_type() != kNoCompression) {
-    block->value = new Block(std::move(contents), raw_block->global_seqno(),
+  if (raw_block_comp_type != kNoCompression) {
+    block->value = new Block(std::move(uncompressed_block_contents), seq_no,
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: can we rename block to cached_block to avoid confusion?

File: table/block_based_table_reader.cc [hunk 14]
```diff
@@ -1802,37 +1812,52 @@ Status BlockBasedTable::MaybeLoadDataBlockToCache(
...
-      }
+    s = GetDataBlockFromCache(key, ckey, block_cache, block_cache_compressed,
+                              rep, ro, block_entry, compression_dict,
+                              rep->table_options.read_amp_bytes_per_bit,
+                              is_index, get_context);
+  }
 
-      if (s.ok()) {
+  // Can't find the block from the cache. If I/O is allowed, read from the file.
+  if (block_entry->value == nullptr && !no_io) {
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: From here to the end of the file, all changes seems irrelevant to the PR. can we separate them to another PR?

File: table/block_based_table_reader.cc [hunk 15]
```diff
@@ -1872,7 +1897,8 @@ BlockBasedTable::PartitionedIndexIteratorState::NewSecondaryIterator(
     Statistics* kNullStats = nullptr;
     return block->second.value->NewIterator<IndexBlockIter>(
         &rep->internal_comparator, rep->internal_comparator.user_comparator(),
-        nullptr, kNullStats, true, index_key_includes_seq_, index_key_is_full_);
+        nullptr, kNullStats, true, index_key_includes_seq_, index_key_is_full_,
+        false /* block_contents_pinned */);
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: Default of block_contents_pinned is false. can we skip passing false here?

File: table/block_based_table_reader.cc [hunk 16]
```diff
@@ -1230,32 +1235,35 @@ Status BlockBasedTable::GetDataBlockFromCache(
...
   // Insert uncompressed block into block cache
   if (s.ok()) {
     block->value =
-        new Block(std::move(contents), compressed_block->global_seqno(),
+        new Block(std::move(contents), rep->get_global_seqno(is_index),
                   read_amp_bytes_per_bit,
                   statistics);  // uncompressed block
-    assert(block->value->compression_type() == kNoCompression);
-    if (block_cache != nullptr && block->value->cachable() &&
+    if (block_cache != nullptr && block->value->own_bytes() &&
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: There are 3 other block_cache->Insert statements in this file (for index and filters). How come we do not check own_bytes() before inserting them into block cache?

File: table/block.cc (hunk 1 above)
Comment (siying) [reply table/block.cc]: But I need to do it, otherwise the format will be too bad:)

File: table/block_based_table_reader.cc (hunk 9 above)
Comment (siying) [reply table/block_based_table_reader.cc]: Previously, global seqno is calculated when blocks are initially read and stored inside block contents. Now it is not stored and is always generated when Block class is generated, so index becomes related.

**Comment by siying** [comment]
Address comments.

File: table/block_based_table_reader.h [hunk 17]
```diff
@@ -256,13 +256,13 @@ class BlockBasedTable : public TableReader {
...
   //    in uncompressed block cache, also sets cache_handle to reference that
   //    block.
-  static Status MaybeLoadDataBlockToCache(FilePrefetchBuffer* prefetch_buffer,
-                                          Rep* rep, const ReadOptions& ro,
-                                          const BlockHandle& handle,
-                                          Slice compression_dict,
-                                          CachableEntry<Block>* block_entry,
-                                          bool is_index = false,
-                                          GetContext* get_context = nullptr);
+  static Status LoadBlockFromCacheOrFile(FilePrefetchBuffer* prefetch_buffer,
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply table/block_based_table_reader.h]: To me the side-effect of cache being loaded is more important information that the source of the data.

File: table/block_based_table_reader.cc [hunk 18]
```diff
@@ -1696,37 +1709,31 @@ TBlockIter* BlockBasedTable::NewDataBlockIterator(
...
+    // Block contents are pinned and can be the pinned guarantee is transferred
+    // after delegating cleanup function, when
+    // 1. block cache is going to be released in cleanup function, or
+    // 2. it's pointing to immortable source. We can determine this case by
+    // checking
+    //    (1) the block doesn't own bytes itself (outside source, usually
+    //    mmapped bytes) (2) the source is immortable
+    bool block_contents_pinned =
+        (block.cache_handle != nullptr ||
+         (!block.value->own_bytes() && rep->immortal_table));
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: can you explain more? immortable_table was for mmap files, where we do not own our bytes. I do not understand how being compressed makes a difference here.

File: table/block_based_table_reader.cc (hunk 15 above)
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: I do not understand the response. The comment was saying that since the default is already false, we do not need to pass it explicitly.

File: table/block_based_table_reader.cc (hunk 2 above)
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: Any thoughts on this?

File: table/block_fetcher.cc [hunk 19]
```diff
@@ -183,6 +183,9 @@ void BlockFetcher::GetBlockContents() {
     }
     *contents_ = BlockContents(std::move(heap_buf_), block_size_);
   }
+#ifdef NDEBUG
```
Comment (maysamyabandeh) [inline table/block_fetcher.cc]: should it be ifndef NDEBUG?

File: table/format.h [hunk 20]
```diff
@@ -193,6 +193,10 @@ struct BlockContents {
   Slice data;     // Actual contents of data
   CacheAllocationPtr allocation;
 
+#ifdef NDEBUG
```
Comment (maysamyabandeh) [inline table/format.h]: should it be ifndef NDEBUG?

File: table/block.h [hunk 21]
```diff
@@ -153,14 +153,12 @@ class Block {
 
   size_t size() const { return size_; }
   const char* data() const { return data_; }
-  bool cachable() const { return contents_.cachable; }
   // The additional memory space taken by the block data.
   size_t usable_size() const { return contents_.usable_size(); }
   uint32_t NumRestarts() const;
+  bool own_bytes() const { return contents_.allocation.get() != nullptr; }
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply table/block.h]: any thoughts on this?

File: table/block_based_table_reader.cc (hunk 2 above)
Comment (siying) [reply table/block_based_table_reader.cc]: Hmm, Right I use DeleteCachedEntry

File: table/block_based_table_reader.cc (hunk 3 above)
Comment (siying) [reply table/block_based_table_reader.cc]: What't the benefit do you think if we do that? The code is shorter? I explicitly pass the argument so that people see it's not pinned.
Saying that, I can change if you think that's the proper way. It's not a big deal.

File: table/block_based_table_reader.cc (hunk 7 above)
Comment (siying) [reply table/block_based_table_reader.cc]: See the PR title and summary. I updated it after I added the change.

File: table/block_based_table_reader.cc (hunk 8 above)
Comment (siying) [reply table/block_based_table_reader.cc]: Yes I agree, but UncompressBlockContents() is used in many other places. I don't think it belongs to this PR.

File: table/block_based_table_reader.cc (hunk 16 above)
Comment (siying) [reply table/block_based_table_reader.cc]: It's not new code. It's just a mapping from previous cachable() call. The three other places are for index, filter block and dummy costing.

File: table/block_based_table_reader.cc (hunk 18 above)
[the remark this answers is not in the document]
Comment (siying) [reply table/block_based_table_reader.cc]: If the source is compressed, MaybeLoadDataBlockToCache() will decompressed the block to heap allocated memory and return. And then here we have a allocation handle.

File: table/block_based_table_builder.cc [hunk 22]
```diff
@@ -677,7 +677,7 @@ Status BlockBasedTableBuilder::InsertBlockInCache(const Slice& block_contents,
     memcpy(ubuf.get(), block_contents.data(), size);
     ubuf[size] = type;
 
-    BlockContents results(std::move(ubuf), size, true, type);
```
[the remark this answers is not in the document]
Comment (siying) [reply table/block_based_table_builder.cc]: cachable field in compressed block is never used, so I guess it's a no harm bug.

File: table/block_fetcher.cc (hunk 19 above)
Comment (siying) [reply table/block_fetcher.cc]: It has been fixed?

File: table/format.h (hunk 20 above)
Comment (siying) [reply table/format.h]: It has been fixed?

File: table/block_based_table_reader.cc (hunk 3 above)
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: Your way is better.

File: table/block_based_table_reader.cc (hunk 16 above)
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: I understand it is not new code but as we are doing a refactoring now, it is a good time to understand the old code better. If the the source does not own bytes, like in mmap files, why should we add them to the block cache? Lets say we add an index block from mmap file to block cache and them later the corresponding memory of mmap file is overwritten but we keep reading from it since the block cache refers to it.

File: table/block_based_table_reader.cc (hunk 12 above)
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: can we add a few lines to the PutDataBlockToCache declaration to clarify that the passed raw_block_contents object is undefined after the function call?

File: table/block_based_table_reader.cc [hunk 23]
```diff
@@ -1300,63 +1308,69 @@ Status BlockBasedTable::PutDataBlockToCache(
...
       // Avoid the following code to delete this cached block.
-      raw_block = nullptr;
+      block_cont_for_comp_cache = nullptr;
       RecordTick(statistics, BLOCK_CACHE_COMPRESSED_ADD);
     } else {
       RecordTick(statistics, BLOCK_CACHE_COMPRESSED_ADD_FAILURES);
     }
+    delete block_cont_for_comp_cache;
   }
-  delete raw_block;
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: would be great to clarify that in the PutDataBlockToCache API.

File: table/block_based_table_reader.cc [hunk 24]
```diff
@@ -1300,65 +1312,72 @@ Status BlockBasedTable::PutDataBlockToCache(
...
+    BlockContents* block_cont_for_comp_cache =
+        new BlockContents(std::move(*raw_block_contents));
+    s = block_cache_compressed->Insert(
+        compressed_block_cache_key, block_cont_for_comp_cache,
+        block_cont_for_comp_cache->ApproximateMemoryUsage(),
+        &DeleteCachedEntry<BlockContents>);
     if (s.ok()) {
       // Avoid the following code to delete this cached block.
-      raw_block = nullptr;
+      block_cont_for_comp_cache = nullptr;
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: can we simply call delete block_cont_for_comp_cache when s.ok() is false?

File: table/block_based_table_reader.cc [hunk 25]
```diff
@@ -1696,37 +1709,31 @@ TBlockIter* BlockBasedTable::NewDataBlockIterator(
   } else {
     iter = new TBlockIter;
   }
-  // Didn't get any data from block caches.
+  // Didn't get any data and there isn't any error. Must be no_io = true
```
[the remark this answers is not in the document]
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: It would make the patch much simpler if we do separate them.

File: table/block_based_table_reader.h [hunk 26]
```diff
@@ -534,6 +533,10 @@ struct BlockBasedTable::Rep {
 
   bool closed = false;
   const bool immortal_table;
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.h]: I could not find where we make use of immortal_table after these changes!

File: table/block_based_table_reader.h [hunk 27]
```diff
@@ -300,30 +298,31 @@ class BlockBasedTable : public TableReader {
...
   // Put a raw block (maybe compressed) to the corresponding block caches.
   // This method will perform decompression against raw_block if needed and then
   // populate the block caches.
   // On success, Status::OK will be returned; also @block will be populated with
   // uncompressed block and its cache handle.
   //
-  // REQUIRES: raw_block is heap-allocated. PutDataBlockToCache() will be
-  // responsible for releasing its memory if error occurs.
+  // Allocated memory managed by raw_block_contents will be transferred to
+  // PutDataBlockToCache(). After the call, the class will be invalid.
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.h]: s/class/object/

File: include/rocksdb/table.h [hunk 28]
```diff
@@ -138,6 +138,8 @@ struct BlockBasedTableOptions {
 
   // If non-NULL use the specified cache for compressed blocks.
   // If NULL, rocksdb will not use a compressed block cache.
+  // Note: though it looks similar too `block_cache`, RocksDB doesn't put the
```
Comment (maysamyabandeh) [inline include/rocksdb/table.h]: s/too/to/

File: table/block_based_table_reader.cc [hunk 29]
```diff
@@ -1723,10 +1747,23 @@ TBlockIter* BlockBasedTable::NewDataBlockIterator(
   if (s.ok()) {
     assert(block.value != nullptr);
     const bool kTotalOrderSeek = true;
+    // Block contents are pinned and it is still pinned after the iterator
+    // is destoryed as long as cleanup functions are moved to another object,
+    // when:
+    // 1. block cache handle is set to be released in cleanup function, or
+    // 2. it's pointing to immortable source. If own_bytes is true then we are
+    //    not reading data from the original source, weather immortal or not.
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: nit: do you see extra spaces here too?

File: table/block_based_table_reader.cc [hunk 30]
```diff
@@ -1723,10 +1747,23 @@ TBlockIter* BlockBasedTable::NewDataBlockIterator(
...
+    //    Otherwise, the block is pinned iff the source is immortal.
+    //    We can determine this case by
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: I guess we can delete these three lines as they are redundant. The rationale is already explained well above and these three lines just repeat what exactly the source does.

File: table/block_based_table_reader.cc [hunk 31]
```diff
@@ -1771,25 +1808,28 @@ TBlockIter* BlockBasedTable::NewDataBlockIterator(
...
   Cache* block_cache = rep->table_options.block_cache.get();
+
+  // No point to cache compressed blocks if it never goes away
   Cache* block_cache_compressed =
-      rep->table_options.block_cache_compressed.get();
+      rep->immortal_table ? nullptr
+                          : rep->table_options.block_cache_compressed.get();
 
+  // First, try to get the block from the cache
+  //
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: nit: is this empty comment line extra?

File: table/block_based_table_reader.cc [hunk 32]
```diff
@@ -1802,38 +1842,45 @@ Status BlockBasedTable::MaybeLoadDataBlockToCache(
...
-            compression_dict, rep->persistent_cache_options,
-            is_index ? kDisableGlobalSequenceNumber : rep->global_seqno,
-            rep->table_options.read_amp_bytes_per_bit,
-            GetMemoryAllocator(rep->table_options), rep->immortal_table);
-      }
+    s = GetDataBlockFromCache(key, ckey, block_cache, block_cache_compressed,
+                              rep, ro, block_entry, compression_dict,
+                              rep->table_options.read_amp_bytes_per_bit,
+                              is_index, get_context);
+  }
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: What is that the block is closed here?

File: table/block_based_table_reader.cc [hunk 33]
```diff
@@ -1771,25 +1807,28 @@ TBlockIter* BlockBasedTable::NewDataBlockIterator(
   return iter;
 }
 
-Status BlockBasedTable::MaybeLoadDataBlockToCache(
+Status BlockBasedTable::ReadBlockAndMaybeLoadToCache(
```
Comment (maysamyabandeh) [inline table/block_based_table_reader.cc]: ReadBlockAndMaybeLoadToCache is misleading since we would not read the data if fill_cache is false. I think the old MaybeLoadDataBlockToCache is the most descriptive.

File: table/block_based_table_reader.cc (hunk 31 above)
Comment (siying) [reply table/block_based_table_reader.cc]: This is to distinguish from the next comment line, which applies to fewer statements than this one.

File: table/block_based_table_reader.cc (hunk 33 above)
Comment (siying) [reply table/block_based_table_reader.cc]: How about MaybeReadBlockAndLoadToCache?

**Comment by maysamyabandeh** [comment]
There are a couple of minor comments pending. LGTM otherwise.
0. Can we clarify in this PR why own_bytes is not used when index/filter is inserted into block cache? If we agree it is a bug, we can create an up_for_grab not to lose track of it.
1. Can you comment on travis failures?
2. I would run it under ASAN to check for possible leaks.
3. a db_bench would give higher confidence that we did not mistakenly omit a block cache update,

**Comment by siying** [comment]
ASAN passed and I'm investigating Travis.

File: table/block_based_table_reader.cc (hunk 33 above)
Comment (maysamyabandeh) [reply table/block_based_table_reader.cc]: That is better.

**Comment by siying** [comment]
db_bench doesn't show any difference.
ASAN passed.
I'm still looking at Travis failure. It looks like a legit failure.
